core/schema/container.go

```diff
@@ -456,89 +807,114 @@ func (s *containerSchema) labels(ctx context.Context, parent *core.Container, ar
+	none := dagql.Null[dagql.String]()
+
 	cfg, err := parent.ImageConfig(ctx)
 	if err != nil {
-		return nil, err
+		return none, err
 	}
 
 	if val, ok := cfg.Labels[args.Name]; ok {
-		return &val, nil
+		return dagql.NonNull(dagql.NewString(val)), nil
 	}
 
-	return nil, nil
+	return none, nil
 }
 
 type containerWithMountedDirectoryArgs struct {
 	Path   string
 	Source core.DirectoryID
-	Owner  string
+	Owner  string `default:""`
 }
 
 func (s *containerSchema) withMountedDirectory(ctx context.Context, parent *core.Container, args containerWithMountedDirectoryArgs) (*core.Container, error) {
-	dir, err := args.Source.Decode()
+	dir, err := args.Source.Load(ctx, s.srv)
 	if err != nil {
 		return nil, err
 	}
-	return parent.WithMountedDirectory(ctx, s.bk, args.Path, dir, args.Owner, false)
+	return parent.WithMountedDirectory(ctx, args.Path, dir.Self, args.Owner, false)
 }
 
 type containerPublishArgs struct {
-	Address           string
-	PlatformVariants  []core.ContainerID
-	ForcedCompression core.ImageLayerCompression
-	MediaTypes        core.ImageMediaTypes
+	Address           dagql.String
+	PlatformVariants  []core.ContainerID `default:"[]"`
+	ForcedCompression dagql.Optional[core.ImageLayerCompression]
+	MediaTypes        core.ImageMediaTypes `default:"OCIMediaTypes"`
 }
 
-func (s *containerSchema) publish(ctx context.Context, parent *core.Container, args containerPublishArgs) (string, error) {
-	return parent.Publish(ctx, s.bk, s.svcs, args.Address, args.PlatformVariants, args.ForcedCompression, args.MediaTypes)
+func (s *containerSchema) publish(ctx context.Context, parent *core.Container, args containerPublishArgs) (dagql.String, error) {
+	variants, err := dagql.LoadIDs(ctx, s.srv, args.PlatformVariants)
+	if err != nil {
+		return "", err
+	}
+	ref, err := parent.Publish(
+		ctx,
+		args.Address.String(),
+		variants,
+		args.ForcedCompression.Value,
+		args.MediaTypes,
+	)
+	if err != nil {
+		return "", err
+	}
+	return dagql.NewString(ref), nil
 }
 
 type containerWithMountedFileArgs struct {
 	Path   string
 	Source core.FileID
-	Owner  string
+	Owner  string `default:""`
 }
 
 func (s *containerSchema) withMountedFile(ctx context.Context, parent *core.Container, args containerWithMountedFileArgs) (*core.Container, error) {
-	file, err := args.Source.Decode()
+	file, err := args.Source.Load(ctx, s.srv)
 	if err != nil {
 		return nil, err
 	}
-	return parent.WithMountedFile(ctx, s.bk, args.Path, file, args.Owner, false)
+	return parent.WithMountedFile(ctx, args.Path, file.Self, args.Owner, false)
 }
 
 type containerWithMountedCacheArgs struct {
-	Path        string
-	Cache       core.CacheVolumeID
-	Source      core.DirectoryID
-	Concurrency core.CacheSharingMode
-	Owner       string
+	Path    string
+	Cache   core.CacheVolumeID
+	Source  dagql.Optional[core.DirectoryID]
+	Sharing core.CacheSharingMode `default:"SHARED"`
```

Comment:
Wait, was this previously misnamed `Concurrency`? I don't see we have test coverage for it, so I guess it might have been?

Reply:
...You might be onto something there. The actual schema has it as [`sharing`](https://github.com/dagger/dagger/blob/674feb34bcbfa75ada4377e019a440aabd3fae27/core/schema/container.graphqls#L372) so I think that arg has just been completely ignored this whole time. Which has pretty spooky implications, and explains some vague memories. We probably don't have tests for it because this kind of thing is a pain in the butt to test. :face_exhaling:

Reply:
Well if nothing else this is at least a nice advantage of the new single source of truth approach! <laughcry>